File: test/arg_parser_test.rb
require 'test/unit'
require_relative '../lib/arg_parser'

# Unit tests for ArgParser
class ArgParserTest < Test::Unit::TestCase
  def test_instantiation
    assert_nothing_thrown { ArgParser.new([]) }
  end

  def test_switch_missing
    args = ArgParser.new([])
    assert_nil(args.switch(%w(-s --long)))
  end

  def test_switch_short
    args = ArgParser.new(%w(-s=short))
    assert_equal('short', args.switch(%w(-s --long)))
  end

  def test_switch_long
    args = ArgParser.new(%w(--long=long))
    assert_equal('long', args.switch(%w(-s --long)))
  end

  def test_switch_overriding_short
    args = ArgParser.new(%w(-s=short1 -s=short2))
    assert_equal('short2', args.switch(%w(-s --long)))
  end

  def test_switch_overriding_long
    args = ArgParser.new(%w(--long=long1 --long=long2))
    assert_equal('long2', args.switch(%w(-s --long)))
  end

  def test_switch_overriding_mixed
    args = ArgParser.new(%w(-s=short --long=long))
    assert_equal('short', args.switch(%w(-s --long)))
    assert_equal('long', args.switch(%w(--long -s)))
  end

  def test_switch_with_space
    args = ArgParser.new(['--switch=long switch'])
    assert_equal('long switch', args.switch('switch'))
  end

  def test_flag_missing
    args = ArgParser.new([])
    assert_equal(false, args.flag?('nope'))
  end

  def test_flag
    args = ArgParser.new(%w(--flag))
    assert_equal(true, args.flag?('flag'))
  end

  def test_flags_mixed_with_stray
    args = ArgParser.new(%w(--flag stray))
    assert_equal(true, args.flag?('flag'))
    assert_equal(%w(stray), args.stray)
  end

  def test_stray_missing
    args = ArgParser.new([])
    assert_equal([], args.stray)
  end

  def def_stray
    args = ArgParser.new(%w(stray1 stray2))
    assert_equal(%w(stray1 stray2), args.stray)
  end

  def test_stray_with_space
    args = ArgParser.new(['long stray'])
    assert_equal(['long stray'], args.stray)
  end

  def test_mixed_types
    args = ArgParser.new(%w(stray1 --switch=s --flag1 stray2 --flag2))

    assert_equal(%w(stray1 stray2), args.stray)
    assert_equal(true, args.flag?('flag1'))
    assert_equal(true, args.flag?('flag2'))
    assert_equal('s', args.switch('switch'))
  end
end

File: tests/arg_parser_test.c
#include "arg_parser.h"
#include "assert_ex.h"

void test_switch_missing()
{
    ArgParser *ap = arg_parser_create();
    assert_null(arg_parser_get_switch(ap, "-s"));
    assert_null(arg_parser_get_switch(ap, "--long"));
    assert_that(!arg_parser_has_switch(ap, "-s"));
    assert_that(!arg_parser_has_switch(ap, "--long"));
    arg_parser_destroy(ap);
}

void test_switch_is_not_a_flag()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"-f"};
    arg_parser_parse(ap, 1, argv);
    assert_null(arg_parser_get_switch(ap, "-f"));
    assert_that(!arg_parser_has_switch(ap, "-f"));
    arg_parser_destroy(ap);
}

void test_switch_short()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"-s=short"};
    arg_parser_parse(ap, 1, argv);
    assert_equals("short", arg_parser_get_switch(ap, "-s"));
    assert_equals("short", arg_parser_get_switch(ap, "s"));
    assert_that(arg_parser_has_switch(ap, "-s"));
    assert_that(arg_parser_has_switch(ap, "s"));
    arg_parser_destroy(ap);
}

void test_switch_long()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"--long=long"};
    arg_parser_parse(ap, 1, argv);
    assert_equals("long", arg_parser_get_switch(ap, "--long"));
    assert_equals("long", arg_parser_get_switch(ap, "-long"));
    assert_equals("long", arg_parser_get_switch(ap, "long"));
    assert_that(arg_parser_has_switch(ap, "--long"));
    assert_that(arg_parser_has_switch(ap, "-long"));
    assert_that(arg_parser_has_switch(ap, "long"));
    arg_parser_destroy(ap);
}

void test_switch_overriding_short()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[2] = {"-s=short1", "-s=short2"};
    arg_parser_parse(ap, 2, argv);
    assert_equals("short2", arg_parser_get_switch(ap, "-s"));
    arg_parser_destroy(ap);
}

void test_switch_overriding_long()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[2] = {"--long=long1", "--long=long2"};
    arg_parser_parse(ap, 2, argv);
    assert_equals("long2", arg_parser_get_switch(ap, "--long"));
    arg_parser_destroy(ap);
}

void test_switch_with_space()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"--switch=long switch"};
    arg_parser_parse(ap, 1, argv);
    assert_equals("long switch", arg_parser_get_switch(ap, "--switch"));
    arg_parser_destroy(ap);
}

void test_switch_empty_value()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"--switch="};
    arg_parser_parse(ap, 1, argv);
    assert_equals("", arg_parser_get_switch(ap, "--switch"));
    arg_parser_destroy(ap);
}

void test_flag_missing()
{
    ArgParser *ap = arg_parser_create();
    assert_that(!arg_parser_has_flag(ap, "nope"));
    arg_parser_destroy(ap);
}

void test_flag()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"--flag"};
    arg_parser_parse(ap, 1, argv);
    assert_that(arg_parser_has_flag(ap, "flag"));
    arg_parser_destroy(ap);
}

void test_flag_mixed_with_stray()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[2] = {"--flag", "stray"};
    arg_parser_parse(ap, 2, argv);
    assert_that(arg_parser_has_flag(ap, "flag"));
    Array *stray = arg_parser_get_stray(ap);
    assert_equali(1, array_size(stray));
    assert_equals("stray", (const char*)array_get(stray, 0));
    arg_parser_destroy(ap);
}

void test_stray_missing()
{
    ArgParser *ap = arg_parser_create();
    Array *stray = arg_parser_get_stray(ap);
    assert_equali(0, array_size(stray));
    arg_parser_destroy(ap);
}

void test_stray()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[2] = {"stray1", "stray2"};
    arg_parser_parse(ap, 2, argv);
    Array *stray = arg_parser_get_stray(ap);
    assert_equali(2, array_size(stray));
    assert_equals("stray1", (const char*)array_get(stray, 0));
    assert_equals("stray2", (const char*)array_get(stray, 1));
    arg_parser_destroy(ap);
}

void test_stray_with_space()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[1] = {"long stray"};
    arg_parser_parse(ap, 1, argv);
    Array *stray = arg_parser_get_stray(ap);
    assert_equali(1, array_size(stray));
    assert_equals("long stray", (const char*)array_get(stray, 0));
    arg_parser_destroy(ap);
}

void test_mixed_types()
{
    ArgParser *ap = arg_parser_create();
    const char *argv[5] = {"stray1", "--switch=s", "--flag1", "stray2", "--flag2"};
    arg_parser_parse(ap, 5, argv);

    assert_equals("s", arg_parser_get_switch(ap, "--switch"));
    assert_that(arg_parser_has_flag(ap, "flag1"));
    assert_that(arg_parser_has_flag(ap, "flag2"));

    Array *stray = arg_parser_get_stray(ap);
    assert_equali(2, array_size(stray));
    assert_equals("stray1", (const char*)array_get(stray, 0));
    assert_equals("stray2", (const char*)array_get(stray, 1));

    arg_parser_destroy(ap);
}

#if 0
#endif

int main(void)
{
    test_switch_missing();
    test_switch_is_not_a_flag();
    test_switch_short();
    test_switch_long();
    // TODO: make this pass
    #if 0
    test_switch_overriding_short();
    test_switch_overriding_long();
    #endif
    test_switch_with_space();
    test_switch_empty_value();
    test_flag_missing();
    test_flag();
    test_flag_mixed_with_stray();
    test_stray_missing();
    test_stray();
    test_stray_with_space();
    test_mixed_types();
    return 0;
}
